src/client/pages/ChangePassword.tsx

```diff
@@ -10,15 +18,240 @@ import { PageBox } from '@/client/components/PageBox';
+) => {
+  if (validationResult === PasswordValidationResult.VALID_PASSWORD) {
+    return ValidationStyling.SUCCESS;
+  }
+  if (redError !== undefined) {
+    return ValidationStyling.ERROR;
+  }
+  return ValidationStyling.FAILURE;
+};
+
+const updateComponentStateAfterValidation = (
+  validationResult: PasswordValidationResult,
+  password: string,
+  passwordCurrently: MutableRefObject<string>,
+  setRedError: Dispatch<SetStateAction<ErrorValidationResult | undefined>>,
+  redErrorCurrently: MutableRefObject<ErrorValidationResult | undefined>,
+  setLastLengthError: Dispatch<SetStateAction<LengthValidationResult>>,
+  setValidationResult: Dispatch<SetStateAction<PasswordValidationResult>>,
+  isPasswordConfirmSelectedCurrently: MutableRefObject<boolean>,
+) => {
+  // return early if the password changed before validation completed
+  if (password !== passwordCurrently.current) return;
+
+  // clear the red error if it's solved (common password does not get solved by going beneath 8 characters)
+  if (redErrorCurrently.current !== undefined) {
+    if (
+      validationResult === PasswordValidationResult.VALID_PASSWORD ||
+      (redErrorCurrently.current !== PasswordValidationResult.COMMON_PASSWORD &&
+        validationResult === PasswordValidationResult.COMMON_PASSWORD)
+    ) {
+      setRedError(undefined);
+    }
+  }
+
+  // set red if user has selected the confirm password check box and there are validation issues
+  if (
+    isPasswordConfirmSelectedCurrently.current &&
+    validationResult !== PasswordValidationResult.VALID_PASSWORD
+  ) {
+    setRedError(validationResult);
+  }
+
+  // set the last length error - when the user solves 'at least 8' or 'max 72' we show the length requirement in green
+  if (
+    validationResult === PasswordValidationResult.AT_LEAST_8 ||
+    validationResult === PasswordValidationResult.MAXIMUM_72
+  ) {
+    setLastLengthError(validationResult);
+  }
+
+  setValidationResult(validationResult);
+};
+
+// The UX requirements for password validation are subtly complex
+// see Reset password flow: https://www.figma.com/file/HrS0AJGKiUzV83dHr1UZP3/Reset-password-screen?node-id=160%3A260
+const usePasswordValidationHooks = (idapiBaseUrl: string) => {
+  // useRefState makes the up-to-date password accessible in the breachPasswordCheck promise handler
+  const [password, setPassword, passwordCurrently] = useRefState('');
+  const [passwordConfirm, setPasswordConfirm] = useState('');
+
+  // if breached password promise completes with errors, and the password confirmation box is selected we show a red error message
+  const [
+    ,
+    setIsPasswordConfirmSelected,
+    isPasswordConfirmSelectedCurrently,
+  ] = useRefState(false);
+
+  // the latest validation result
+  const [
+    validationResult,
+    setValidationResult,
+  ] = useState<PasswordValidationResult>(PasswordValidationResult.AT_LEAST_8);
+
+  // redError is used for the first password input box
```

Comment:
I've attempted to describe the some of the UX behaviour, and the reasoning behind the state. I will attach a video to the PR to make it more clear.